File: mol2db2/mol2db2.py
#!/usr/bin/env python

#Ryan G. Coleman

import string
import sys
import optparse
import sybyl2dock
import mol2
import hierarchy
import solv
import clash
import hydrogens
import time
import math

from hierarchy import TooBigError

def mol2db2(options):
  '''function that does all the actual work you may want to do to convert a
  mol2 file and solv file into a db2 file.'''
  if options.timeit:
    timeStart = time.time()
  else:
    timeStart = None
  mol2data = mol2.Mol2(options.mol2file, nameFileName=options.namefile)
  mol2data.convertDockTypes(options.atomtypefile)
  mol2data.addColors(options.colortablefile)
  solvdata = solv.Solv(options.solvfile)
  if options.covalent:
    covAtomType,indicesList = mol2data.removeCovalentDummyAtom()
    mol2data.recolorCovalentAttachment(covAtomType)
    for index in indicesList:
      del solvdata.charge[index]
      del solvdata.polarSolv[index]
      del solvdata.surface[index]
      del solvdata.apolarSolv[index]
      del solvdata.solv[index]

  if options.verbose:
    print ("names:", mol2data.name, mol2data.protName, mol2data.smiles, \
        mol2data.longname)
    print ("dock atom types:", mol2data.atomType)
    print ("dock atom type numbers:", mol2data.dockNum)
    print ("dock color type numbers:", mol2data.colorNum)
  clashDecider = clash.Clash(options.clashfile)
  hydrogenRotater = hydrogens.Hydrogens(options.hydrogenfile)
  if options.timeit:
    timeReadIn = time.time()
    print ("time to read in files:", timeReadIn-timeStart)
  #0th step is to do the hydrogen operation directly on the mol2data.
  #done here so the hierarchy knows exactly how many conformations it must deal
  #with, not some estimate. and so each hierarchy has the max # of allowed sets
  #again, without guessing.
  if options.rotateh or options.reseth:
    hydrogenRotater.findTerminalHydrogens(mol2data)
    if options.verbose:
      print (mol2data.hydrogensToRotate, " hydrogens need rotated")
    if options.reseth and 0 < mol2data.hydrogensToRotate:
      hydrogenRotater.resetHydrogens(mol2data)
    if options.rotateh and 0 < mol2data.hydrogensToRotate:
      mol2data = hydrogenRotater.rotateHydrogens(mol2data)
  if options.timeit:
    hydTime = time.time()
    print ("time to move hydrogens:", hydTime-timeReadIn)
  if options.verbose:
    print (len(mol2data.atomXyz), " conformations in input")

  def hierarchyDataGenerator(this_mol2data, depth=1):
    '''Generators to pipeline hierarchy generation'''
    try:
      yield hierarchy.Hierarchy(
          this_mol2data, clashDecider, tolerance=options.tolerance,
          verbose=options.verbose, timeit=options.timeit,
          limitset=options.limitset, limitconf=options.limitconf,
          limitcoord=options.limitcoord, solvdata=solvdata)
    except TooBigError as limitError:
      if depth > options.maxrecursiondepth:
        raise
      breaks = hierarchy.computeBreaks(limitError, options)
      origConfsPer = max(len(this_mol2data.atomXyz)/(breaks + 1), 1)  # at least 1
      if options.verbose:
        print ("splitting original input conformations into", breaks + 1, "parts", "(current depth: ", depth, ')')
      for snap in range(breaks + 1):  # extra one to do leftovers
        newMol2data = this_mol2data.copy()  # copy orig before hydrogen rotations
        first = origConfsPer * snap
        last = origConfsPer * (snap + 1)
        newMol2data.keepConfsOnly(first, last)
        if len(newMol2data.atomXyz) > 0:
          subgen = hierarchyDataGenerator(newMol2data, depth=depth+1)
          for subhier in subgen:
              yield subhier
  if options.timeit:
    timeHier = time.time()
    print ("time to (start) construction hierarchy (subtotal):", timeHier-hydTime)
  return timeStart, hierarchyDataGenerator(mol2data)

def mol2db2writeDb2(options, timeStart, hierarchyDatas):
  '''does the writing of the output files. separate so you can make but
  not write. requires timeStart (can be None if no times wanted) and
  all the hierarchyDatas from the mol2db2 function'''
  if options.timeit:
    timeHier = time.time()
  # Clear db2gzfile
  open(options.db2gzfile, 'wb').close()
  for hierarchyData in hierarchyDatas:
    hierarchyData.write(
        db2gzFileName=options.db2gzfile, verbose=options.verbose,
        timeit=options.timeit, limitset=options.limitset,
        writeMode='a')  # append so we don't overwrite
    yield hierarchyData
  if options.timeit:
    timeHierOut = time.time()
    print ("time to save db2:", timeHierOut-timeHier)
  if options.timeit:
    timeFinal = time.time()
    print ("time to do everything (total):", timeFinal-timeStart)

def parserDefaults():
  '''uses optparse to make an options parser. useful here to allow calling by
  other python modules without command line interface.'''
  description = "Convert multimol2 and solv files into db2 files"
  usage = "%prog [options]"
  parser = optparse.OptionParser(usage=usage, description=description)
  parser.add_option(
      "-v", "--verbose", action="store_true", dest="verbose", default=False,
      help="lots of debugging output")
  parser.add_option(
      "--covalent", action="store_true", dest="covalent", default=False,
      help="make a DOCKovalent compatible db (assumes SiH2 dummy atom)")
  parser.add_option(
      "-a", "--time", action="store_true", dest="timeit", default=False,
      help="timing output")
  parser.add_option(
      "-m", "--mol2", type="string", action="store", dest="mol2file",
      default="db.mol2.gz", help="multimol2 input file, (default: %default)")
  parser.add_option(
      "-n", "--namemol2", type="string", action="store", dest="namefile",
      default="name.txt", help="mol2 name file, (default: %default)")
  parser.add_option(
      "-s", "--solv", type="string", action="store", dest="solvfile",
      default="db.solv", help="solv input file, (default: %default)")
  parser.add_option(
      "-o", "--db", type="string", action="store", dest="db2gzfile",
      default="db.db2.gz", help="db2.gz output file, (default: %default)")
  parser.add_option(
      "-t", "--atomtype", type="string", action="store", dest="atomtypefile",
      default=None, help="atom type conversion file, (default: %default)")
  parser.add_option(
      "-c", "--colortable", type="string", action="store",
      dest="colortablefile",
      default=None, help="color table conversion file, (default: %default)")
  parser.add_option(
      "-d", "--clash", type="string", action="store", dest="clashfile",
      default=None, help="clash checking parameter file, (default: %default)")
  parser.add_option(
      "-y", "--hydrogens", type="string", action="store", dest="hydrogenfile",
      default=None,
      help="terminal hydrogen parameter file, (default: %default)")
  parser.add_option(
      "-r", "--noreseth", action="store_false", dest="reseth", default=True,
      help="reset the planar terminal hydrogen positions, (default: %default)")
  parser.add_option(
      "-z", "--norotateh", action="store_false", dest="rotateh", default=True,
      help="rotate the terminal hydrogen positions, (default: %default)")
  #yes you can do reset and rotate in any combination!
  parser.add_option(
      "-x", "--disttol", type="float", action="store", dest="tolerance",
      default=0.001,
      help="distance tolerance in angstroms, (default: %default)")
  parser.add_option(
      "-l", "--limitset", type="long", action="store", dest="limitset",
      default=2500000, help="limit on the number of sets, (default: %default)")
  parser.add_option(
      "--limitconf", type="long", action="store", dest="limitconf",
      default=200000, help="limit on the number of confs, (default: %default)")
  parser.add_option(
      "--limitcoord", type="long", action="store", dest="limitcoord",
      default=1000000, help="limit on the number of coords, (default: %default)")
  parser.add_option(
      "--maxrecursiondepth", type=int, action="store", dest="maxrecursiondepth",
      default=1, help="Max recursive subdivision steps to take (default: %default)")
  return parser

if -1 != sys.argv[0].find("mol2db2.py") and __name__ == '__main__':
  parser = parserDefaults()
  options, args = parser.parse_args()  # default reads from argv[1:]
  if 0 != len(args):
    parser.error("mol2db2.py takes no positional arguments" +
                 "  Use --help for more information")
  else:
    #print options
    if options.verbose:
      print ("verbose debugging requested.")
      print (options)
    timeStart, hierarchyDatas = mol2db2(options)  # main program call
    finishedHierarchyDatas = mol2db2writeDb2(
        options, timeStart, hierarchyDatas)  # write output
    for hierarchyData in finishedHierarchyDatas:
      if options.verbose:
        print ("Cleaning up")
      del hierarchyData

File: mol2db2_cpp_strain/src/main/mol2db2.cpp
#include "mol2.h"
#include "clash.h"
#include "hydrogens.h"
#include "hierarchy.h"
#include <iostream>
#include <string>
#include <cstring>

using namespace mol2db2;

void printUsage(const char* progName) {
    std::cout << "Usage: " << progName << " [options]\n";
    std::cout << "\nOptions:\n";
    std::cout << "  -v, --verbose           Verbose output\n";
    std::cout << "  --covalent              Process covalent ligands (remove SiH3 dummy)\n";
    std::cout << "  -a, --time              Show timing information\n";
    std::cout << "  -m, --mol2 FILE         Input mol2 file (default: db.mol2.gz)\n";
    std::cout << "  -n, --namemol2 FILE     Name file (default: name.txt)\n";
    std::cout << "  -s, --solv FILE         Solvation file (default: db.solv)\n";
    std::cout << "  -o, --db FILE           Output db2.gz file (default: db.db2.gz)\n";
    std::cout << "  -t, --atomtype FILE     Atom type conversion file\n";
    std::cout << "  -c, --colortable FILE   Color table file\n";
    std::cout << "  -d, --clash FILE        Clash parameter file\n";
    std::cout << "  -y, --hydrogens FILE    Hydrogen parameter file\n";
    std::cout << "  -r, --noreseth          Don't reset planar hydrogens\n";
    std::cout << "  -z, --norotateh         Don't rotate terminal hydrogens\n";
    std::cout << "  -x, --disttol FLOAT     Distance tolerance (default: 0.001)\n";
    std::cout << "  --limitset NUM          Maximum number of sets (default: 2500000)\n";
    std::cout << "  --limitconf NUM         Maximum number of conformations (default: 200000)\n";
    std::cout << "  --limitcoord NUM        Maximum number of coordinates (default: 1000000)\n";
    std::cout << "  --help                  Show this help message\n";
}

int main(int argc, char* argv[]) {
    /* Default options */
    bool verbose = false;
    bool covalent = false;
    bool timeit = false;
    std::string mol2file = "db.mol2.gz";
    std::string namefile = "name.txt";
    std::string solvfile = "db.solv";
    std::string db2gzfile = "db.db2.gz";
    std::string* atomtypefile = nullptr;
    std::string* colortablefile = nullptr;
    std::string* clashfile = nullptr;
    std::string* hydrogenfile = nullptr;
    bool reseth = true;
    bool rotateh = true;
    double tolerance = 0.001;
    long limitset = 2500000;
    long limitconf = 200000;
    long limitcoord = 1000000;
    
    /* Parse command line arguments */
    for (int i = 1; i < argc; i++) {
        std::string arg = argv[i];
        
        if (arg == "-v" || arg == "--verbose") {
            verbose = true;
        } else if (arg == "--covalent") {
            covalent = true;
        } else if (arg == "-a" || arg == "--time") {
            timeit = true;
        } else if (arg == "--help") {
            printUsage(argv[0]);
            return 0;
        } else if (arg == "-m" || arg == "--mol2") {
            if (i + 1 < argc) {
                mol2file = argv[++i];
            }
        } else if (arg == "-n" || arg == "--namemol2") {
            if (i + 1 < argc) {
                namefile = argv[++i];
            }
        } else if (arg == "-s" || arg == "--solv") {
            if (i + 1 < argc) {
                solvfile = argv[++i];
            }
        } else if (arg == "-o" || arg == "--db") {
            if (i + 1 < argc) {
                db2gzfile = argv[++i];
            }
        } else if (arg == "-t" || arg == "--atomtype") {
            if (i + 1 < argc) {
                atomtypefile = new std::string(argv[++i]);
            }
        } else if (arg == "-c" || arg == "--colortable") {
            if (i + 1 < argc) {
                colortablefile = new std::string(argv[++i]);
            }
        } else if (arg == "-d" || arg == "--clash") {
            if (i + 1 < argc) {
                clashfile = new std::string(argv[++i]);
            }
        } else if (arg == "-y" || arg == "--hydrogens") {
            if (i + 1 < argc) {
                hydrogenfile = new std::string(argv[++i]);
            }
        } else if (arg == "-r" || arg == "--noreseth") {
            reseth = false;
        } else if (arg == "-z" || arg == "--norotateh") {
            rotateh = false;
        } else if (arg == "-x" || arg == "--disttol") {
            if (i + 1 < argc) {
                tolerance = std::stod(argv[++i]);
            }
        } else if (arg == "--limitset") {
            if (i + 1 < argc) {
                limitset = std::stol(argv[++i]);
            }
        } else if (arg == "--limitconf") {
            if (i + 1 < argc) {
                limitconf = std::stol(argv[++i]);
            }
        } else if (arg == "--limitcoord") {
            if (i + 1 < argc) {
                limitcoord = std::stol(argv[++i]);
            }
        } else {
            std::cerr << "Unknown option: " << arg << std::endl;
            printUsage(argv[0]);
            return 1;
        }
    }
    
    if (verbose) {
        std::cout << "mol2db2 - Converting MOL2 to DB2 format\n";
        std::cout << "Input MOL2: " << mol2file << "\n";
        std::cout << "Input SOLV: " << solvfile << "\n";
        std::cout << "Output DB2: " << db2gzfile << "\n";
    }
    
    try {
        /* Read mol2 file */
        if (verbose) {
            std::cout << "Reading MOL2 file..." << std::endl;
        }
        Mol2 mol2data(&mol2file, &namefile, nullptr);
        
        /* Convert dock types and add colors */
        if (verbose) {
            std::cout << "Converting atom types..." << std::endl;
        }
        mol2data.convertDockTypes(atomtypefile);
        mol2data.addColors(colortablefile);
        
        /* Read solvation data */
        if (verbose) {
            std::cout << "Reading solvation data..." << std::endl;
        }
        mol2data.solvData.readSolvFile(solvfile);
        
        /* Handle covalent ligands */
        if (covalent) {
            if (verbose) {
                std::cout << "Processing covalent ligand..." << std::endl;
            }
            auto result = mol2data.removeCovalentDummyAtom();
            std::string covAtomType = std::get<0>(result);
            std::vector<int> indicesList = std::get<1>(result);
            
            mol2data.recolorCovalentAttachment(covAtomType);
            
            /* Remove corresponding solv entries */
            for (int idx : indicesList) {
                if (idx < (int)mol2data.solvData.charge.size()) {
                    mol2data.solvData.charge.erase(
                        mol2data.solvData.charge.begin() + idx);
                    mol2data.solvData.polarSolv.erase(
                        mol2data.solvData.polarSolv.begin() + idx);
                    mol2data.solvData.surface.erase(
                        mol2data.solvData.surface.begin() + idx);
                    mol2data.solvData.apolarSolv.erase(
                        mol2data.solvData.apolarSolv.begin() + idx);
                    mol2data.solvData.solv.erase(
                        mol2data.solvData.solv.begin() + idx);
                }
            }
        }
        
        /* Create clash decider */
        Clash clashDecider(clashfile);
        
        /* Handle hydrogens */
        Hydrogens hydrogenRotater(hydrogenfile);
        
        if (rotateh || reseth) {
            if (verbose) {
                std::cout << "Processing terminal hydrogens..." << std::endl;
            }
            
            hydrogenRotater.findTerminalHydrogens(&mol2data);
            
            if (verbose) {
                std::cout << "Found " << mol2data.hydrogensToRotate 
                         << " rotatable hydrogens" << std::endl;
            }
            
            if (reseth && mol2data.hydrogensToRotate > 0) {
                if (verbose) {
                    std::cout << "Resetting planar hydrogens..." << std::endl;
                }
                hydrogenRotater.resetHydrogens(&mol2data);
            }
            
            if (rotateh && mol2data.hydrogensToRotate > 0) {
                if (verbose) {
                    std::cout << "Rotating hydrogens..." << std::endl;
                }
                //mol2data = hydrogenRotater.rotateHydrogens(mol2data);
                hydrogenRotater.rotateHydrogens_teb(mol2data);
            }
        }
        
        if (verbose) {
            std::cout << "Total conformations: " << mol2data.atomXyz.size() 
                     << std::endl;
        }
        
        /* Build hierarchy */
        if (verbose) {
            std::cout << "Building hierarchy..." << std::endl;
        }
        
        Hierarchy hierarchy(mol2data, clashDecider, tolerance, verbose,
                            timeit, limitset, limitconf, limitcoord);
        
        /* Write output */
        if (verbose) {
            std::cout << "Writing DB2 file..." << std::endl;
        }
        
        hierarchy.write(db2gzfile, verbose, timeit, limitset, "w");
        
        if (verbose) {
            std::cout << "Conversion complete!" << std::endl;
        }
        
        /* Clean up */
        delete atomtypefile;
        delete colortablefile;
        delete clashfile;
        delete hydrogenfile;
        
        return 0;
        
    } catch (const TooBigError& e) {
        std::cerr << "Error: Hierarchy too large!\n";
        std::cerr << "  Conformations: " << e.getConfs() << "\n";
        std::cerr << "  Sets: " << e.getSets() << "\n";
        std::cerr << "  Coordinates: " << e.getCoords() << "\n";
        std::cerr << "Try adjusting --limitset, --limitconf, or --limitcoord\n";
        return 1;
    } catch (const std::exception& e) {
        std::cerr << "Error: " << e.what() << std::endl;
        return 1;
    }
}
